wandb/sdk/wandb_settings.py

```diff
@@ -394,11 +412,54 @@ def infer_settings_from_env(self):
         # For code saving, only allow env var override if value from server is true, or
         # if no preference was specified.
         if (self.save_code is True or self.save_code is None) and os.getenv(
-            env.SAVE_CODE
+            wandb.env.SAVE_CODE
         ) is not None:
-            u["save_code"] = env.should_save_code()
+            u["save_code"] = wandb.env.should_save_code()
+
+        if self.jupyter:
+            meta = wandb.jupyter.notebook_metadata()
+            u["_jupyter_path"] = meta.get("path")
+            u["_jupyter_name"] = meta.get("name")
+            u["_jupyter_root"] = meta.get("root")
+
+        # host and username are populated by env_settings above if their env
+        # vars exist -- but if they don't, we'll fill them in here
+        if not self.host:
+            u["host"] = socket.gethostname()
+
+        if not self.username:
+            try:
+                u["username"] = getpass.getuser()
+            except KeyError:
+                # getuser() could raise KeyError in restricted environments like
+                # chroot jails or docker containers.  Return user id in these cases.
+                u["username"] = str(os.getuid())
+
+        u["_executable"] = sys.executable
+
+        u["docker"] = wandb.env.get_docker()
+
+        try:
+            pynvml.nvmlInit()
+            u["_gpu"] = pynvml.nvmlDeviceGetName(
+                pynvml.nvmlDeviceGetHandleByIndex(0)
+            ).decode("utf8")
+            u["_gpu_count"] = pynvml.nvmlDeviceGetCount()
+        except pynvml.NVMLError:
+            pass
+        try:
+            u["_cpu_count"] = multiprocessing.cpu_count()
+        except NotImplementedError:
+            pass
```

Comment:
Lets keep the _gpu* and _cpu* probing in the internal process in meta like you had it.  Since that is where we are doing similar stuff for stats.   Then we always know if they are using pynvml it is in one process context.